### __Predicting Used Car Sale Prices__

--- 
**Overview:**   

- The goal is to predict prices of used cars based on car characteristics and regression models
- Assessment Summary:  
    - Write a problem statement and perform Exploratory Data Analysis  
    - Clean up data, deal with categorical features and missing observations, and create new explanatory variables (feature engineering)  
    - Construct and tune forecasting models, produce forecasts and submit your predictions 


**Marks**: 

- Total Marks: 40
- Your individual mark will consist of:  
    - 50% x overall assignment mark + 45% x mark for the task that you are responsible for + 5% x mark received from your teammates for your effort in group work 
- 1 mark: Ranking in the top 5 places of your unit on Kaggle  
- 3 marks: Reaching the first place in the competition   


**Submissions:**  

1. On Kaggle: submit your team's forecast in order to be ranked by Kaggle
    - Limit of 20 submission per day
2. On iLearn **only team leader to submit** this Jupyter notebook re-named `Group_Assignment_Team_Name.ipynb` where Team_Name is your team's name on Kaggle   
    - The Jupyter notebook must contain team members names/ID numbers, and team name in the competition
    - Provide answers to the 3 Tasks below in the allocated cells including all codes/outputs/writeups 
    - One 15 minute video recording of your work 
        - Each team member to provide a 5 minute presentation of the Task that they led (it is best to jointly record your video using Zoom)
        - When recording your video make sure your face is visible, that you share your Jupyter Notebook and explain everything you've done in the submitted Jupyter notebook on screen
        - 5 marks will be deducted from each Task for which there is no video presentation or if you don't follow the above instructions
        
3. On iLearn each student needs to submit a file with their teammates' names, ID number and a mark for their group effort (out of 100%)

---

**Fill out the following information**

For each team member provide name, Student ID number and which task is performed below

- Team Name on Kaggle: `BUSA8001_NAVIKS`
- Team Leader and Team Member 1: `Rohit, 47811684`
- Team Member 2: `Nav Mahajan, 47738642`
- Team Member 3: `Vikas Chahal, 47694998`

---

## Task 1: Problem Description and Initial Data Analysis

1. Read the Competition Overview on Kaggle [https://www.kaggle.com/t/32b34f072642495487836cf93453ac6a](https://www.kaggle.com/t/32b34f072642495487836cf93453ac6a)
2. Referring to Competition Overview and the data provided on Kaggle write **Problem Description** (about 500 words) focusing on key points that will need to be addressed as first steps in Tasks 2 and 3 below, 

- Using the following headings:
    - Forecasting Problem - explain what you are trying to do and how it could be used in the real world (i.e. why it may be important)
    - Evaluation Criteria - explain the criteria is used to assess forecast performance 
    - Types of Variables/Features
    - Data summary and main data characteristics
    - Missing Values (only explain what you find - do not impute missing values at this stage)
    - You should **not** discuss any specific predictive algorithms at this stage
    - Note: Your written portion of this task should be completed in a single Markdown cell
    
Total Marks: 12


In [1]:
#Loading necessary Libraries

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline 

In [2]:
#Loading Training Data

df_train = pd.read_csv("train.csv")

In [3]:
#Getting Information of each variable

df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 39 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   vin                   3500 non-null   object 
 1   back_legroom          3397 non-null   object 
 2   body_type             3494 non-null   object 
 3   city                  3500 non-null   object 
 4   city_fuel_economy     2912 non-null   float64
 5   daysonmarket          3500 non-null   int64  
 6   dealer_zip            3500 non-null   int64  
 7   engine_displacement   3375 non-null   float64
 8   engine_type           3450 non-null   object 
 9   exterior_color        3500 non-null   object 
 10  franchise_dealer      3500 non-null   bool   
 11  front_legroom         3397 non-null   object 
 12  fuel_tank_volume      3397 non-null   object 
 13  fuel_type             3463 non-null   object 
 14  height                3397 non-null   object 
 15  highway_fuel_economy 

In [4]:
#Checking null values in the train dataset

df_train.isnull().sum()


vin                       0
back_legroom            103
body_type                 6
city                      0
city_fuel_economy       588
daysonmarket              0
dealer_zip                0
engine_displacement     125
engine_type              50
exterior_color            0
franchise_dealer          0
front_legroom           103
fuel_tank_volume        103
fuel_type                37
height                  103
highway_fuel_economy    588
horsepower              125
interior_color            0
is_new                    0
latitude                  0
length                  103
listed_date               0
listing_color             0
longitude                 0
make_name                 0
maximum_seating         103
mileage                 203
model_name                0
power                   299
savings_amount            0
seller_rating             0
torque                  331
transmission             60
transmission_display     60
wheel_system            101
wheelbase           

In [5]:
# Loading test dataset

df_test = pd.read_csv("test.csv")

In [6]:
# #Getting Information of each variable

df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1500 entries, 0 to 1499
Data columns (total 38 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   vin                   1500 non-null   object 
 1   back_legroom          1447 non-null   object 
 2   body_type             1498 non-null   object 
 3   city                  1500 non-null   object 
 4   city_fuel_economy     1263 non-null   float64
 5   daysonmarket          1500 non-null   int64  
 6   dealer_zip            1500 non-null   int64  
 7   engine_displacement   1441 non-null   float64
 8   engine_type           1464 non-null   object 
 9   exterior_color        1500 non-null   object 
 10  franchise_dealer      1500 non-null   bool   
 11  front_legroom         1447 non-null   object 
 12  fuel_tank_volume      1447 non-null   object 
 13  fuel_type             1471 non-null   object 
 14  height                1447 non-null   object 
 15  highway_fuel_economy 

In [7]:
#Checking null values in the test dataset

df_test.isnull().sum()

vin                       0
back_legroom             53
body_type                 2
city                      0
city_fuel_economy       237
daysonmarket              0
dealer_zip                0
engine_displacement      59
engine_type              36
exterior_color            0
franchise_dealer          0
front_legroom            53
fuel_tank_volume         53
fuel_type                29
height                   53
highway_fuel_economy    237
horsepower               59
interior_color            0
is_new                    0
latitude                  0
length                   53
listed_date               0
listing_color             0
longitude                 0
make_name                 0
maximum_seating          53
mileage                  76
model_name                0
power                   126
savings_amount            0
seller_rating             3
torque                  141
transmission             13
transmission_display     13
wheel_system             43
wheelbase           

In [8]:
df_train.head().transpose()

0                       1  \
vin                       SJKCH5CRXHA032566       5LMCJ3D96HUL54638   
back_legroom                        33.5 in                 36.8 in   
body_type                   SUV / Crossover         SUV / Crossover   
city                             Great Neck                   Wayne   
city_fuel_economy                       NaN                    19.0   
daysonmarket                             20                      64   
dealer_zip                            11021                    7470   
engine_displacement                  2000.0                  2300.0   
engine_type                              I4                      I4   
exterior_color                        Black  Magnetic Gray Metallic   
franchise_dealer                      False                    True   
front_legroom                       41.3 in                 42.8 in   
fuel_tank_volume                   14.8 gal                15.5 gal   
fuel_type                          Gasoline                Gasoline   
height                              60.2 in                 65.6 in   
highway_fuel_economy                    NaN                    25.0   
horsepower                            208.0                   275.0   
interior_color                  Brown (Tan)              Cappuccino   
is_new                                False                   False   
latitude                            40.7774                 40.9086   
length                             174.2 in                179.2 in   
listed_date                        20-08-20                07-07-20   
listing_color                         BLACK                    GRAY   
longitude                          -73.7268                -74.2609   
make_name                          INFINITI                 Lincoln   
maximum_seating                     5 seats                 5 seats   
mileage                             40186.0                 53874.0   
model_name                             QX30                     MKC   
power                    208 hp @ 5,500 RPM      275 hp @ 5,500 RPM   
savings_amount                         1522                     740   
seller_rating                      3.346154                    4.75   
torque                258 lb-ft @ 1,200 RPM   300 lb-ft @ 2,000 RPM   
transmission                              A                       A   
transmission_display      7-Speed Automatic       6-Speed Automatic   
wheel_system                            AWD                     AWD   
wheelbase                          106.3 in                105.9 in   
width                                 82 in                 84.1 in   
year                                   2017                    2017   
price                                 21495                   22459   

                                                              2  \
vin                                           5LMCJ2D95HUL35217   
back_legroom                                            36.8 in   
body_type                                       SUV / Crossover   
city                                                      Wayne   
city_fuel_economy                                          19.0   
daysonmarket                                                 14   
dealer_zip                                                 7470   
engine_displacement                                      2300.0   
engine_type                                                  I4   
exterior_color        Burgundy Velvet Metallic Tinted Clearcoat   
franchise_dealer                                           True   
front_legroom                                           42.8 in   
fuel_tank_volume                                       15.5 gal   
fuel_type                                              Gasoline   
height                                                  65.6 in   
highway_fuel_economy                                       25.0   
horsepower                                                275.

In [11]:
#Checking summary stats of Train Dataset

df_train.describe().transpose()

count          mean           std        min  \
city_fuel_economy     2912.0     21.655220      7.520448    10.0000   
daysonmarket          3500.0     78.987714    104.838545     0.0000   
dealer_zip            3500.0  11876.966857  10893.589741   922.0000   
engine_displacement   3375.0   2749.185185   1115.578544  1000.0000   
highway_fuel_economy  2912.0     28.658997      7.160677    13.0000   
horsepower            3375.0    242.296889     87.287417    78.0000   
latitude              3500.0     40.549032      3.426925    18.3988   
longitude             3500.0    -74.212195      2.947159   -84.3466   
mileage               3297.0  39277.213224  47300.958932     0.0000   
savings_amount        3500.0    830.168286   1318.130267     0.0000   
seller_rating         3500.0      3.934368      0.542599     1.0000   
year                  3500.0   2017.015714      4.037270  1978.0000   
price                 3500.0  27133.806000  16502.784211   899.0000   

                               25%           50%         75%          max  
city_fuel_economy        18.000000     21.000000     24.0000     127.0000  
daysonmarket             17.000000     39.500000     82.0000     698.0000  
dealer_zip             7036.000000   8812.000000  11706.0000   49203.0000  
engine_displacement    2000.000000   2400.000000   3500.0000    7000.0000  
highway_fuel_economy     25.000000     28.000000     31.0000     122.0000  
horsepower              176.000000    240.000000    292.0000     797.0000  
latitude                 40.745700     40.855800     41.0912      42.7851  
longitude               -74.075800    -73.726800    -73.2587     -66.0785  
mileage                  14.000000  25983.000000  57778.0000  329145.0000  
savings_amount            0.000000    206.000000   1255.7500   12801.0000  
seller_rating             3.577778      4.115385      4.3750       5.0000  
year                   2016.000000   2018.000000   2020.0000    2021.0000  
price                 15995.000000  25488.000000  34012.5000  207900.0000

`(Task 1, Text Here - insert more cells as required)`

# Task1

# Problem Description:

## Forecasting Problem:

We are tasked with developing a predictive model capable of forecasting the prices of used cars based on a comprehensive set of vehicle attributes. This forecasting model addresses a crucial question in the automotive industry: how can we accurately determine the value of a used car? By leveraging historical data, including make, model, mileage, condition, and additional features, we aim to predict the selling price for each car listed in the dataset.

In the real world, this forecasting capability is invaluable for a range of stakeholders:

### Consumers:
It assists potential buyers in determining whether a used car is priced fairly, helping them make informed purchasing decisions.

### Dealerships: 
It aids car dealers in setting competitive and profitable pricing for their inventory.

### Insurance Companies: 
It helps insurers in assessing vehicle value for policy writing and claims handling.

### Market Analysts:
It provides insights into market trends, helping analysts predict future market dynamics based on historical sales data.

By accurately forecasting car prices, we can facilitate a more transparent, efficient, and fair marketplace, benefiting both sellers and buyers by aligning expectations and reducing transaction friction.

## Evaluation Criteria
To assess the performance of our forecasting models, we will employ a set of rigorous evaluation criteria designed to measure the accuracy and reliability of the predicted prices. The primary metrics will be:

### Root Mean Squared Error (RMSE): 
This metric evaluates the average magnitude of the errors between the predicted and actual prices, with a lower RMSE indicating better model performance.

## Types of Variables/Features:

### Numerical Columns:

back_legroom
city_fuel_economy
daysonmarket
dealer_zip
height
front_legroom 
fuel_tank_volume 
engine_displacement
highway_fuel_economy
horsepower
length 
latitude
longitude
mileage
savings_amount
power 
torque 
seller_rating
wheelbase 
width 
year
price


### Categorical Columns:

vin
body_type
city
engine_type
exterior_color
franchise_dealer 
fuel_type
interior_color
is_new 
listed_date
listing_color
make_name
maximum_seating
model_name
transmission
transmission_display
wheel_system


## Missing Data:

### For Training Data:
Complete Data: Several columns have no missing values, indicating complete information across all 3500 entries. These include identifiers like vin, location details such as city, dealer_zip, latitude, longitude, and vehicle details like exterior_color, interior_color, make_name, model_name, and others.

Fuel Economy Data: The variables city_fuel_economy and highway_fuel_economy have 588 missing entries each, which could significantly impact fuel economy-related predictions or insights.

Dimensional Attributes: Measurements such as back_legroom, front_legroom, fuel_tank_volume, height, length, maximum_seating, wheelbase, and width each have 103 missing values. These missing entries will need to be addressed, possibly through imputation, as they could affect predictions related to vehicle size and space.

Engine and Performance Data: engine_displacement and horsepower have 125 missing values each, and torque has 331 missing values. These are critical features for predicting vehicle performance and price.

Miscellaneous Attributes: A few other attributes have a small number of missing values, such as body_type (6), engine_type (50), fuel_type (37), mileage (203), power (299), transmission (60), and transmission_display (60). wheel_system has 101 missing entries.

Zero Missing Data: Key variables for vehicle transaction details like daysonmarket, savings_amount, seller_rating, year, and price have no missing values. This is crucial since price is the target variable for prediction and others could be strong predictors.

### For Test Data:

Complete Variables: Many attributes have no missing entries, ensuring that these features can be reliably used in the model evaluation phase. This includes identifying information (vin), location (city, dealer_zip, latitude, longitude), vehicle specifics (exterior_color, interior_color, make_name, model_name, listed_date, listing_color), and transaction details (daysonmarket, is_new, savings_amount, year).

Dimensional and Comfort Attributes: There are missing values in features that describe the physical dimensions and comfort of the vehicles, such as back_legroom, front_legroom, fuel_tank_volume, height, length, maximum_seating, wheelbase, and width, each with 53 missing entries.

Fuel Economy: Both city_fuel_economy and highway_fuel_economy have 237 missing values, significant for any fuel-related analysis or predictions.

Engine Specifications: Critical performance-related features like engine_displacement, horsepower, and torque have 59, 59, and 141 missing values, respectively, which could impact the model's accuracy in predicting vehicle performance or value.

Miscellaneous Attributes: Minor missing data are present in body_type (2), engine_type (36), fuel_type (29), mileage (76), power (126), transmission (13), transmission_display (13), wheel_system (43), and a small number in seller_rating (3).

## Data summary and main data characteristics:
:
city_fuel_economy: Most cars have a city fuel economy averaging around 21.66 miles per gallon with a standard deviation of 7.52, indicating some variability. The range is quite wide, going from 10 to 127 miles per gallon, which suggests that there are outliers or potentially errors in the data, as 127 mpg is unusually high for city driving conditions.
daysonmarket: Cars have been on the market for an average of roughly 79 days.
dealer_zip, longitude and latitude: With a large range in zip codes from 922 to 49203 indicates dealers are spread across different regions.
engine_displacement: With a range from 1000 cc to 7000 cc, indicating a mix of small to large engine sizes.
highway_fuel_economy: The average highway fuel economy is better than city fuel economy, at around 28.66 mpg.
horsepower: The average horsepower is 242, with a wide range from 78 to 797, suggesting a diverse set of cars from low-power to high-performance vehicles.
mileage: The high standard deviation indicates a wide variety of car conditions.
seller_rating: The average seller rating is close to 4 out of 5, with a relatively small standard deviation, suggesting that most sellers are rated fairly well.
year: The cars range from vintage 1978 models to nearly new 2021 models, with an average year of 2017. This suggests that most cars in the dataset are relatively recent.
price: With a very wide range from $899 to $207,900, indicating a dataset that includes a wide range of car values, from economical to luxury models.

**Main Data Characteristics:**
Skewness and Outliers: The presence of outliers is suggested by the max values in variables like city_fuel_economy, highway_fuel_economy, and price. These need to be investigated to ensure they are not data entry errors.

Variability: Some variables show high variability (as indicated by the standard deviation), such as daysonmarket, savings_amount, and price. This suggests that the data covers a wide range of scenarios.

Data Distribution: Several variables might not be normally distributed, indicated by the difference between the mean and median values. For instance, daysonmarket, savings_amount, and price show right-skewed distributions since the mean is higher than the median.

Age of Cars: The year variable indicates a relatively modern set of cars, but the range shows that some vintage cars are included.




---

## Task 2: Data Cleaning, Missing Observations and Feature Engineering
- In this task you will follow a set of instructions/questions listed below.
- Make sure you explain each answer carefully both in Markdown text and on your video.

Total Marks: 12

**Task 2, Question 1**: Clean **all** numerical features so that they can be used in training algorithms. For instance, back_legroom feature is in object format containing both numerical values and text. Extract numerical values (equivalently eliminate the text) so that the numerical values can be used as a regular feature.  
(2 marks)

In [15]:
# Extracting numerical values

# Training Data split

col_to_split = ["back_legroom", "front_legroom", "fuel_tank_volume", "height", "length", "wheelbase", "width"]

def ext_num_col(text):
    num_part = ''.join(filter(str.isdigit, str(text)))
    if num_part:
        return float(num_part)
    else:
        return None
    
for column in col_to_split:
    df_train[column] = df_train[column].apply(ext_num_col)
    
# test data spliting
for column in col_to_split:
    df_test[column] = df_test[column].apply(ext_num_col)

The columns we chose ["back_legroom", "front_legroom", "fuel_tank_volume", "height", "length", "wheelbase", "width"] were bject types with numerical values in the front and text at the back. We used the above code to extract the numerical values out from both the training and test dataset

`(Task 2, Question 1 Text Here - insert more cells as required)`

**Task 2, Question 2** Create at least 5 new features from the existing numerical variables which contain multiple items of information, for example you could extract maximum torque and torque rpm from the torque variable.  
(2 marks)

In [16]:
df_train[["max_power", "power_rpm"]] = df_train["power"].str.split("@", expand=True)
df_train[["max_torque", "torque_rpm"]] = df_train["torque"].str.split("@", expand=True)


new_col = ["max_power", "power_rpm", "max_torque", "torque_rpm"]

for column in new_col:
    df_train[column] = df_train[column].apply(ext_num_col)
    
df_train["number_of_gears"] = df_train["transmission_display"].str.extract(r'(\d+)')
df_train["transmission_type"] = df_train["transmission_display"].str.replace(r'\d+','').str.strip()

    
df_train.drop(columns=['power', 'torque', 'transmission_display', 'transmission_type'], inplace=True)



C:\Users\nav pc\AppData\Local\Temp\ipykernel_12188\1293162912.py:11: FutureWarning: The default value of regex will change from True to False in a future version.
  df_train["transmission_type"] = df_train["transmission_display"].str.replace(r'\d+','').str.strip()


In the above code, we created 6 new variables from the 3 composite variables in the training dataset which were 'power', 'torque' and 'transmission_display'. We splitted the 'power' to 'max_power'and 'power_rpm', 'torque' to 'max_torque' and 'torque_rpm' and 'transmission_display' to 'number_of_gears' and 'transmission_type'. Atlast we dropped the variables 'power', 'torque', 'transmission_display', 'transmission_type' which were not required.

In [17]:
# for test data

df_test[["max_power", "power_rpm"]] = df_test["power"].str.split("@", expand=True)
df_test[["max_torque", "torque_rpm"]] = df_test["torque"].str.split("@", expand=True)


new_col = ["max_power", "power_rpm", "max_torque", "torque_rpm"]

for column in new_col:
    df_test[column] = df_test[column].apply(ext_num_col)
    
df_test["number_of_gears"] = df_test["transmission_display"].str.extract(r'(\d+)')
df_test["transmission_type"] = df_test["transmission_display"].str.replace(r'\d+','').str.strip()

    
df_test.drop(columns=['power', 'torque', 'transmission_display', 'transmission_type'], inplace=True)


C:\Users\nav pc\AppData\Local\Temp\ipykernel_12188\1161955810.py:13: FutureWarning: The default value of regex will change from True to False in a future version.
  df_test["transmission_type"] = df_test["transmission_display"].str.replace(r'\d+','').str.strip()


In the above code, we created 6 new variables from the 3 composite variables in the test dataset which were 'power', 'torque' and 'transmission_display'. We splitted the 'power' to 'max_power'and 'power_rpm', 'torque' to 'max_torque' and 'torque_rpm' and 'transmission_display' to 'number_of_gears' and 'transmission_type'. Atlast we dropped the variables 'power', 'torque', 'transmission_display', 'transmission_type' which were not required.

**Task 2, Question 3**: Impute missing values for all features in both the training and test datasets.   
(3 marks)

In [22]:
#Imputing Missing values

numerical_columns = ['back_legroom', 'city_fuel_economy', 'engine_displacement', 'front_legroom', 
                     'fuel_tank_volume', 'height', 'highway_fuel_economy', 'horsepower', 'length', 
                     'mileage', 'wheelbase', 'width', 'power_rpm', 'max_power', 'torque_rpm', 'max_torque']
categorical_columns = ['body_type', 'engine_type', 'fuel_type', 'maximum_seating', 'transmission', 'wheel_system']

numerical_mean = df_train[numerical_columns].mean()
categorical_mode = df_train[categorical_columns].mode().iloc[0]

# Add missing values using the mean for numerical columns and the mode for categorical columns
df_train[numerical_columns] = df_train[numerical_columns].fillna(numerical_mean)
df_train[categorical_columns] = df_train[categorical_columns].fillna(categorical_mode)

df_train['number_of_gears'].fillna(0 , inplace = True)
df_train['number_of_gears'] = df_train['number_of_gears'].astype(int)

# Displaying the DataFrame info
print(df_train.info())



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 41 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   vin                   3500 non-null   object 
 1   back_legroom          3500 non-null   float64
 2   body_type             3500 non-null   object 
 3   city                  3500 non-null   object 
 4   city_fuel_economy     3500 non-null   float64
 5   daysonmarket          3500 non-null   int64  
 6   dealer_zip            3500 non-null   int64  
 7   engine_displacement   3500 non-null   float64
 8   engine_type           3500 non-null   object 
 9   exterior_color        3500 non-null   object 
 10  franchise_dealer      3500 non-null   bool   
 11  front_legroom         3500 non-null   float64
 12  fuel_tank_volume      3500 non-null   float64
 13  fuel_type             3500 non-null   object 
 14  height                3500 non-null   float64
 15  highway_fuel_economy 

As we checked for the missing values in task 1, now in this task, we'll be imputing the missing values by first categorizing them into numerical and categorical variables based on their data types. Once the categorization is done, we'll replace the missing values in numerical variables by 'mean' of the columns and the categorical variables by 'mode' of the columns. This is first done on the training dataset.

In [23]:
# For test data 
numerical_columns_test = ['back_legroom', 'city_fuel_economy', 'engine_displacement', 'front_legroom', 
                     'fuel_tank_volume', 'height', 'highway_fuel_economy', 'horsepower', 'length', 
                     'mileage','seller_rating', 'wheelbase', 'width', 'power_rpm', 'max_power', 'torque_rpm', 'max_torque']
categorical_columns_test = ['body_type', 'engine_type', 'fuel_type', 'maximum_seating', 'transmission', 'wheel_system']

numerical_mean_test = df_test[numerical_columns_test].mean()
categorical_mode_test = df_test[categorical_columns_test].mode().iloc[0]

# Add missing values using the mean for numerical columns and the mode for categorical columns
df_test[numerical_columns_test] = df_test[numerical_columns_test].fillna(numerical_mean_test)
df_test[categorical_columns_test] = df_test[categorical_columns_test].fillna(categorical_mode_test)

df_test['number_of_gears'].fillna(0 , inplace = True)
df_test['number_of_gears'] = df_test['number_of_gears'].astype(int)

# Displaying the DataFrame info
print(df_test.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1500 entries, 0 to 1499
Data columns (total 40 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   vin                   1500 non-null   object 
 1   back_legroom          1500 non-null   float64
 2   body_type             1500 non-null   object 
 3   city                  1500 non-null   object 
 4   city_fuel_economy     1500 non-null   float64
 5   daysonmarket          1500 non-null   int64  
 6   dealer_zip            1500 non-null   int64  
 7   engine_displacement   1500 non-null   float64
 8   engine_type           1500 non-null   object 
 9   exterior_color        1500 non-null   object 
 10  franchise_dealer      1500 non-null   bool   
 11  front_legroom         1500 non-null   float64
 12  fuel_tank_volume      1500 non-null   float64
 13  fuel_type             1500 non-null   object 
 14  height                1500 non-null   float64
 15  highway_fuel_economy 

Similarly, we did the above task for test dataset as well.

`(Task 2, Question 3 Text Here - insert more cells as required)`

**Task 2, Question 4**: Encode all categorical variables appropriately as discussed in class. 

- Where multiple values are given for an observation encode the observation as 'other'. 
- Where a categorical feature contains more than 5 unique values, map the features into 5 most frequent values + 'other' and then encode appropriately. For instance, map colours into 5 basic colours + 'other': [red, yellow, green, blue, purple, other] and then encode.  
(2 marks)

In [24]:
#Checking the data for the encoding purpose

categorical_columns_f = ['body_type', 'engine_type', 'fuel_type', 'maximum_seating', 'transmission', 'wheel_system', 'listing_color']
    
for col in categorical_columns_f:
    
    print(f"Category in {col} is : {df_train[col].value_counts()}")
    

Category in body_type is : SUV / Crossover    1910
Sedan              1032
Pickup Truck        166
Coupe                92
Hatchback            88
Minivan              84
Wagon                69
Convertible          34
Van                  25
Name: body_type, dtype: int64
Category in engine_type is : I4                      2093
V6                       803
V8                       267
I6                        82
I4 Hybrid                 63
H4                        45
V8 Flex Fuel Vehicle      35
V6 Flex Fuel Vehicle      35
I3                        25
I5                        15
I6 Diesel                 11
I4 Flex Fuel Vehicle       8
V8 Biodiesel               6
I4 Diesel                  4
V6 Hybrid                  2
V6 Diesel                  2
W12                        1
V8 Diesel                  1
V10                        1
V12                        1
Name: engine_type, dtype: int64
Category in fuel_type is : Gasoline             3320
Flex Fuel Vehicle      78
Hybrid 

We put all the categorical columns of the training dataset into ' categorical_columns_f ' and then checked for the variation in the dataset.

In [25]:
# Using labelEncoder to label the categorical columns

from sklearn.preprocessing import LabelEncoder

def encode_categorical(column):
    label_encoder = LabelEncoder()
    value_counts = column.value_counts()
    top_values = value_counts.index[:5]  # Get the top 5 values
    # Create a mapping for top 5 values
    mapping = {value: label for label, value in enumerate(top_values)}
    # Encode values based on the mapping, label others as 5
    column = column.map(mapping).fillna(5)
    return column

# Encode categorical columns as specified
for column in df_train.columns:
    if column in ['body_type', 'engine_type', 'fuel_type', 'maximum_seating', 'transmission', 'wheel_system', 'listing_color']:
        df_train[column] = encode_categorical(df_train[column])

#df_train.head().transpose()


Using the LabelEncoder,we categorised the catgorical variables into categories starting from 0, wherein 0 is the maximum number and 4 is the least referring to the question. All the remaining data is encoded to 5.

In [26]:
# Encode categorical columns as specified for test data
for column in df_test.columns:
    if column in ['body_type', 'engine_type', 'fuel_type', 'maximum_seating', 'transmission', 'wheel_system', 'listing_color']:
        df_test[column] = encode_categorical(df_test[column])

Performing the above same task on the test dataset as well.

`(Task 2, Question 4 Text Here - insert more cells as required)`

**Task 2, Question 5**: Perform any other actions you think need to be done on the data before constructing predictive models, and clearly explain what you have done.   
(1 marks)

In [30]:
def encode_boolean(column): 
    label_encoder = LabelEncoder() 
    column = label_encoder.fit_transform(column) 
    return column 

# Encode boolean columns 

df_train["franchise_dealer"] = encode_boolean(df_train["franchise_dealer"]) 
df_train["is_new"] = encode_boolean(df_train["is_new"])

df_train.head().transpose()

0                       1  \
vin                   SJKCH5CRXHA032566       5LMCJ3D96HUL54638   
back_legroom                      335.0                   368.0   
body_type                           0.0                     0.0   
city                         Great Neck                   Wayne   
city_fuel_economy              21.65522                    19.0   
daysonmarket                         20                      64   
dealer_zip                        11021                    7470   
engine_displacement              2000.0                  2300.0   
engine_type                         0.0                     0.0   
exterior_color                    Black  Magnetic Gray Metallic   
franchise_dealer                      0                       1   
front_legroom                     413.0                   428.0   
fuel_tank_volume                  148.0                   155.0   
fuel_type                           0.0                     0.0   
height                            602.0                   656.0   
highway_fuel_economy          28.658997                    25.0   
horsepower                        208.0                   275.0   
interior_color              Brown (Tan)              Cappuccino   
is_new                                0                       0   
latitude                        40.7774                 40.9086   
length                           1742.0                  1792.0   
listed_date                    20-08-20                07-07-20   
listing_color                       0.0                     2.0   
longitude                      -73.7268                -74.2609   
make_name                      INFINITI                 Lincoln   
maximum_seating                     0.0                     0.0   
mileage                         40186.0                 53874.0   
model_name                         QX30                     MKC   
savings_amount                     1522                     740   
seller_rating                  3.346154                    4.75   
transmission                          0                       0   
wheel_system                          0                       0   
wheelbase                        1063.0                  1059.0   
width                              82.0                   841.0   
year                               2017                    2017   
price                             21495                   22459   
max_power                         208.0                   275.0   
power_rpm                        5500.0                  5500.0   
max_torque                        258.0                   300.0   
torque_rpm                       1200.0                  2000.0   
number_of_gears                       7                       6   

                                                              2  \
vin                                           5LMCJ2D95HUL35217   
back_legroom                                              368.0   
body_type                                                   0.0   
city                                                      Wayne   
city_fuel_economy                                          19.0   
daysonmarket                                                 14   
dealer_zip                                                 7470   
engine_displacement                                      2300.0   
engine_type                                                 0.0   
exterior_color        Burgundy Velvet Metallic Tinted Clearcoat   
franchise_dealer                                              1   
front_legroom                                             428.0   
fuel_tank_volume                                          155.0   
fuel_type                                                   0.0   
height                                                    656.0   
highway_fuel_economy                                       25.0   
horsepower                                                275.0   
interior_color   

In this task, we could check that the variables "franchise_dealer" and "is_new" were in boolean, so we used the LabelEncoder to convert the true and false to 1 and 0.

In [31]:
# Encode boolean columns for test data

df_test["franchise_dealer"] = encode_boolean(df_test["franchise_dealer"]) 
df_test["is_new"] = encode_boolean(df_test["is_new"])

df_test.head().transpose()

0                         1  \
vin                   5N1AT2MV8HC824461         1GNSKCKC2HR160472   
back_legroom                      379.0                      39.0   
body_type                           0.0                       0.0   
city                        Wallingford          North Plainfield   
city_fuel_economy                  25.0                      16.0   
daysonmarket                         19                        34   
dealer_zip                         6492                      7060   
engine_displacement              2500.0                    5300.0   
engine_type                         0.0                       4.0   
exterior_color             Gun Metallic  Iridescent Pearl Tricoat   
franchise_dealer                      0                         0   
front_legroom                      43.0                     453.0   
fuel_tank_volume                  145.0                      26.0   
fuel_type                           0.0                       1.0   
height                            684.0                     744.0   
highway_fuel_economy               32.0                      22.0   
horsepower                        170.0                     355.0   
interior_color                     None                     Brown   
is_new                                0                         0   
latitude                        41.4772                   40.6229   
length                           1845.0                     204.0   
listed_date                  2020-08-21                2020-08-06   
listing_color                       3.0                       4.0   
longitude                      -72.8124                  -74.4454   
make_name                        Nissan                 Chevrolet   
maximum_seating                     1.0                       2.0   
mileage                         30396.0                   57901.0   
model_name                        Rogue                     Tahoe   
savings_amount                      861                        43   
seller_rating                  4.545455                  4.418605   
transmission                          0                         0   
wheel_system                          0                         2   
wheelbase                        1065.0                     116.0   
width                             724.0                     805.0   
year                               2017                      2017   
max_power                         170.0                     355.0   
power_rpm                        6000.0                    5600.0   
max_torque                        175.0                     383.0   
torque_rpm                       4400.0                    4100.0   
number_of_gears                       0                         6   

                                       2                  3                  4  
vin                    5NPD84LFXJH361029  5XXGT4L34KG284916  2G1WF52E819291457  
back_legroom                       357.0              356.0         345.809355  
body_type                            1.0                1.0                1.0  
city                          West Nyack              Bronx       Little Ferry  
city_fuel_economy                   28.0               24.0               19.0  
daysonmarket                          49                 20                 38  
dealer_zip                         10994              10466               7643  
engine_displacement               2000.0             2400.0             3400.0  
engine_type                          0.0                0.0                1.0  
exterior_color        Quartz White Pearl   Snow White Pearl           Burgundy  
franchise_dealer                       1                  1                  0  
front_legroom                      422.0              455.0         375.583972  
fuel_tank_volume                    14.0              185.0         135.183829  
fuel_type                            0.0                0.0               

The above same task is done on the test dataset as well.

`(Task 2, Question 5 Text Here - insert more cells as required)`

**Task 2, Question 6**: Perform exploratory data analysis to measure the relationship between the features and the target and carefully write up your findings. 
(2 marks)

In [33]:
df_train.corr()

back_legroom  body_type  city_fuel_economy  \
back_legroom          1.000000e+00  -0.121489          -0.012801   
body_type            -1.214888e-01   1.000000           0.069134   
city_fuel_economy    -1.280062e-02   0.069134           1.000000   
daysonmarket         -1.250120e-02  -0.037734           0.047027   
dealer_zip            2.534921e-02   0.035910           0.008601   
engine_displacement   4.927592e-03   0.125925          -0.392370   
engine_type          -5.485596e-02   0.220671          -0.071800   
franchise_dealer      1.650983e-03  -0.162498           0.126030   
front_legroom        -7.690485e-02   0.035101           0.009198   
fuel_tank_volume      1.812422e-01  -0.265865          -0.078360   
fuel_type            -4.622490e-07   0.135804           0.455473   
height               -2.508240e-02   0.025980          -0.077292   
highway_fuel_economy -4.448377e-02   0.131781           0.937369   
horsepower           -9.010204e-02   0.071296          -0.411079   
is_new               -4.242558e-02  -0.176438           0.136048   
latitude              1.759604e-02   0.073488           0.005656   
length                3.278858e-02   0.016002          -0.056164   
listing_color        -3.137393e-02   0.005775           0.047133   
longitude            -2.263338e-02  -0.061433          -0.000417   
maximum_seating      -9.463784e-02   0.345351          -0.216614   
mileage               1.868852e-02   0.150830          -0.149762   
savings_amount        2.654422e-02   0.051004          -0.128241   
seller_rating        -8.011713e-02   0.036497          -0.001183   
transmission         -3.037669e-02   0.100352           0.200080   
wheel_system          2.977110e-02   0.253296          -0.062032   
wheelbase             5.659559e-02   0.050400           0.010138   
width                -4.144725e-02   0.071644          -0.095493   
year                  3.059666e-02  -0.206526           0.154878   
price                -1.146524e-01  -0.144290          -0.129662   
max_power            -7.871835e-02   0.059707          -0.371336   
power_rpm             3.290301e-02   0.000727           0.005217   
max_torque           -5.100832e-02   0.062315          -0.302007   
torque_rpm            1.001926e-01   0.064055          -0.038768   
number_of_gears       2.008350e-02  -0.026165          -0.133902   

                      daysonmarket  dealer_zip  engine_displacement  \
back_legroom             -0.012501    0.025349             0.004928   
body_type                -0.037734    0.035910             0.125925   
city_fuel_economy         0.047027    0.008601            -0.392370   
daysonmarket              1.000000   -0.072963            -0.098197   
dealer_zip               -0.072963    1.000000             0.089186   
engine_displacement      -0.098197    0.089186             1.000000   
engine_type              -0.073608    0.120371             0.468722   
franchise_dealer          0.093371    0.013383            -0.167893   
front_legroom             0.052360    0.071572            -0.059437   
fuel_tank_volume          0.015061   -0.095506            -0.118640   
fuel_type                -0.015305    0.106861             0.121099   
height                   -0.080909    0.056684             0.130286   
highway_fuel_economy      0.039576    0.001000            -0.420912   
horsepower               -0.066211    0.050976             0.760449   
is_new                    0.293144   -0.014961            -0.227466   
latitude                 -0.257613    0.258086             0.055257   
length                   -0.062632    0.103397             0.142319   
listing_color            -0.007261   -0.000627            -0.119290   
longitude                 0.158507   -0.923079            -0.100574   
maximum_seating          -0.050127    0.071585             0.493955   
mileage                  -0.188364   -0.028203             0.172194   
savings_amount           -0.057825    0.003263             0.234080   
seller_r

- The above code is used to find the correlation between all the numerical variables. The correlation matrix clearly states that the back_legroom, body_type, and city_fuel_economy show weak negative correlations with the price, indicating that more legroom, certain body types, or better city fuel efficiency may correspond to slightly lower prices. 

- daysonmarket and franchise_dealer have weak positive correlations with price, suggesting that cars listed for longer or sold by franchise dealers might be priced a bit higher. 

- engine_displacement, horsepower, year, max_power, and max_torque all have moderate to strong positive correlations with price, indicating that larger, more powerful, and newer cars tend to be more expensive. 

- mileage has a strong negative correlation with price, showing that higher mileage significantly lowers the vehicle's price. 

- dealer_zip has a negligible correlation with price, suggesting location has little impact on price in this dataset.

`(Task 2, Question 6 Text Here - insert more cells as required)`

--- 
## Task 3: Fit and tune a forecasting model/Submit predictions/Report score and ranking

Make sure you **clearly explain each step** you do both in text and on the recoded video.   
This task must not create any additional features and has to use on the dataset constructed in Task 2.

1. Build at least 3 machine learning (ML) regression models taking into account the outcomes of Tasks 1 & 2 (Explain Carefully)
2. Fit the models and tune hyperparameters via cross-validation: make sure you comment and explain each step clearly
3. Select your best algorithm, create predictions using the test dataset, and submit your predictions on Kaggle's competition page
4. Provide Kaggle ranking and **score** (screenshot your best submission) and comment
5. Make sure your Python code works, so that a marker that can replicate your all Kaggle Score   

- Hint: to perform well you will need to iterate Tasks 2 and Task 3.

Total Marks: 12

In [34]:
# Importing `StandardScaler` from scikit-learn to prepare for feature scaling.
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

# Defining a list of column names representing numerical data in the `df_train` dataframe.
numerical_col_scaling = ['back_legroom', 'city_fuel_economy', 'engine_displacement', 'front_legroom', 
                         'fuel_tank_volume', 'height', 'highway_fuel_economy', 'horsepower', 'length', 
                         'mileage', 'wheelbase', 'width', 'power_rpm', 'max_power', 'torque_rpm', 'max_torque',
                         'daysonmarket','savings_amount','seller_rating']

df_train[numerical_col_scaling] = scaler.fit_transform(df_train[numerical_col_scaling])


In [35]:
# scaling of test data

df_test[numerical_col_scaling] = scaler.fit_transform(df_test[numerical_col_scaling])


**Scaling of Data**

This standardization process involves subtracting the mean and dividing by the standard deviation for each feature, effectively transforming the data such that each feature has a mean of zero and a standard deviation of one.
The fit_transform method is called on the scaler object with the specified numerical columns of df_train. The fit_transform method does two things:

- fit: It computes the mean and standard deviation of each feature.
- transform: It scales each feature to have zero mean and a standard deviation of one.
The resulting scaled features are then reassigned back to their respective columns in the df_train dataframe.

# Model formation 


## Model 1 - LR

In [36]:

# importing the necessary libraries 

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


# Define the input features and target variable
input_columns = ['engine_displacement', 'franchise_dealer', 'horsepower', 'is_new', 'year', 'max_power', 
                 'torque_rpm', 'max_torque', 'mileage']
target_column = ['price']

# Split the data into train and test sets
X_train = df_train[input_columns]
y_train = df_train[target_column]
X_test = df_test[input_columns]


# Create a Linear Regression model
model = LinearRegression()

# Fit the model to the training data
model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = model.predict(X_test)

# storing the data into csv file
df_test_lr = df_test
df_test_lr['price'] = y_pred
df_lr_1 = df_test_lr[['vin','price']]
df_lr_1.to_csv('df_lr_1.csv', index=False)

The linear regression model was chosen for this task due to its efficiency in modeling linear relationships between independent variables and the target variable, which is the price in this dataset. Linear regression is advantageous for its interpretability—each coefficient of an input feature gives insights into the relationship between that feature and the target variable.

From a technical standpoint, linear regression assumes a linear relationship between the input variables (such as engine displacement, horsepower, and mileage) and the continuous target variable (price). It is suitable here because it's a fundamental algorithm that can serve as a benchmark for more complex models.

By fitting the linear regression model to the training data, we establish a mathematical equation that weights each of the input features according to their relationship with the vehicle's price. This allows us to predict the price based on the input features of new, unseen data, with the expectation that these predictions will be as close as possible to the actual market prices.

The final output of the model—vehicle prices predicted for the test dataset—is stored in a CSV file for practical use. This output can serve multiple purposes, such as setting prices for new inventory, guiding negotiations, or powering an online recommendation system.

**rmse score: 8752.45**
The linear regression model has the highest RMSE, indicating the least predictive accuracy of the three models for this particular dataset. This suggests that the relationship between the predictors and the target variable is not well captured by a linear model, which is consistent with real-world data often being non-linear and complex.


## Model 2 - DT

In [37]:
from sklearn.tree import DecisionTreeRegressor


model = DecisionTreeRegressor(random_state=0)

# Fit the model to the training data
model.fit(X_train, y_train)

# Make predictions on the test data
y_pred_dt = model.predict(X_test)

df_test_dt = df_test
df_test_dt['price'] = y_pred_dt
df_dt_1 = df_test_dt[['vin','price']]
df_dt_1.to_csv('df_dt_1.csv', index=False)

The Decision Tree Regressor was selected for its capability to model complex, non-linear relationships between the features and the target variable, which in this case is the price. This model is particularly useful for capturing the intricate patterns that linear models may miss, providing potentially more accurate predictions for varied datasets.

Unlike linear regression, a decision tree does not assume a straight-line relationship between the input variables (like car's age, brand, mileage, etc.) and the continuous target variable (price). Instead, it breaks down the dataset into smaller subsets while at the same time an associated decision tree is incrementally developed. The result is a model that can capture non-linear relationships in the data.

By training the Decision Tree Regressor on the given data, we create a series of decision rules that can be used to predict the price of a car. It is a more flexible model compared to linear regression, capable of adjusting to more complex scenarios. However, it still maintains a level of interpretability—by examining the paths through the tree, we can understand how the model arrives at its predictions.

The predictions made by the decision tree for the test data are then combined with the unique identifiers for each vehicle, the VINs, and saved into a CSV file. This file translates the model's predictions into a practical format that can be used for various applications, such as informing the pricing strategy for a fleet of cars, providing estimates on a car sales website, or aiding in market analysis.

**rmse score: 8189.25**
The decision tree has a higher RMSE compared to the stacking model, indicating less predictive accuracy. Decision trees are often prone to overfitting to the training data, which can result in poorer performance on unseen data, reflected in a higher RMSE.


# Model 3 Stacking

In [39]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import StackingRegressor, GradientBoostingRegressor

in_col = ['back_legroom', 'body_type', 'city_fuel_economy','daysonmarket','engine_displacement',
          'franchise_dealer', 'front_legroom', 'fuel_tank_volume', 'highway_fuel_economy',
          'horsepower', 'is_new', 'latitude', 'listing_color', 'longitude', 'maximum_seating', 'mileage', 'savings_amount',
          'transmission', 'wheel_system', 'width', 'year', 'max_power', 'power_rpm',
          'max_torque', 'torque_rpm', 'number_of_gears']
targ_col = ['price']

# Split the data into train and test sets
X_train_1 = df_train[in_col]
y_train_1 = df_train[targ_col]
X_test_1 = df_test[in_col]

# Define base models
base_models = [
    ('random_forest', RandomForestRegressor(n_estimators=100, random_state=42)),
    ('gradient_boosting', GradientBoostingRegressor(n_estimators=200, max_depth=4, learning_rate=0.1, loss='squared_error'))
]

# Define the meta-learner model
meta_learner = LinearRegression()

# Create the stacking ensemble
model = StackingRegressor(estimators=base_models, final_estimator=meta_learner)

# ravelleing is used to convert the target variable to a one-dimensional array 

y_train_1 = y_train_1.values.ravel() 
# Fit the stacking model on the entire training dataset
model.fit(X_train_1, y_train_1)

# Make predictions on the test data
y_pred_stack = model.predict(X_test_1)

df_test_rf_3 = df_test
df_test_rf_3['price'] = y_pred_stack
df_rf_3 = df_test_rf_3[['vin','price']]
df_rf_3.to_csv('df_rf_3_1.csv', index=False)



Stacking is powerful because it combines the predictive power of multiple models, which can often result in better performance than any single model alone. It leverages the diversity of different models to improve predictions: one model's weaknesses can be compensated for by another's strengths. The final ensemble, therefore, can potentially provide more accurate and robust predictions.

**Stacking Ensemble**
The StackingRegressor combines the base models and uses the meta-learner to finalize the predictions. This ensemble model works in two steps:

The base models are trained on the data, and their predictions are used as input features for the meta-learner.
The meta-learner is trained on these predictions to make the final prediction.
Preparing the Target Variable
Lastly, y_train_1 = y_train_1.values.ravel() is a line of code ensuring that y_train_1 (which is the target variable you want to predict) is in the correct format for model training. The ravel() function is used to convert the target variable to a one-dimensional array, which is the required format for scikit-learn's fit method.

## Base Models
There are two base models in:

**Random Forest Regressor:** This is an ensemble learning method itself, which operates by constructing multiple decision trees during training and outputting the mean prediction of the individual trees. It is initialized with 100 trees (n_estimators=100) and a random_state for reproducibility. This model operates by building a 'forest' of decision trees, each contributing its insights. The ensemble's final prediction is made by averaging the individual trees' predictions, which tends to give a balanced outcome that is less likely to be swayed by outliers or noise in the dataset.

**Gradient Boosting Regressor:** Another ensemble model that builds trees one at a time, where each new tree helps to correct errors made by previously trained trees. It's initialized with 200 sequential trees (n_estimators=200), a maximum depth of 4 for each tree (max_depth=4), a learning rate of 0.1 to control the contribution of each tree, and it uses squared_error as the loss function to measure how well the model fits the data. 

## Meta-Learner Model
These base models are then orchestrated by a Linear Regression meta-learner, which acts as the conductor, learning the best way to blend the base models' predictions to arrive at a final estimate for the car prices. This meta-learner adds a level of sophistication by weighting the input from each base model according to its effectiveness, much like combining individual expert opinions into a consensus forecast. We are using a simple Linear Regression model for this task. 

**rmse: 5445.38**

This is the lowest RMSE among the three models, suggesting that the stacking model has the best predictive accuracy.. This performance can be attributed to the stacking model's ability to leverage the strengths of multiple predictive models, thereby providing a more nuanced and accurate prediction. It is particularly effective in scenarios where no single model captures all the patterns in the data, whereas a combination of models can achieve a better overall prediction.

## Marking Criteria

- Marking Rubrics
    - Problem Description - 12 marks
    - Data Cleaning - 12 marks
    - Building Forecasting models - 12 marks
    - Competition Points - 4 marks


- To receive full marks your solutions must satisfy the following criteria:
    - Provide Python solutions that follow the modelling methodology developed in BUSA8001   
    - Written answers explain your logic and Python code in detail, and beformulated in easy to understand full sentences   
    
---
---